src/silo/query_engine/actions/aggregated.cpp

```diff
@@ -0,0 +1,304 @@
+   std::vector<OperatorResult> bitmap_filters
+) const {
+   if (group_by_fields.empty()) {
+      uint32_t count = 0;
+      for (const auto& filter : bitmap_filters) {
+         count += filter->cardinality();
+      };
+      std::map<std::string, std::variant<std::string, uint32_t>> tuple_fields;
+      tuple_fields["count"] = count;
+      return QueryResult{std::vector<AggregationResult>{{tuple_fields}}};
+   }
+   if (group_by_fields.size() == 1) {
+      const std::string group_by_field = group_by_fields[0];
+      const auto& metadata = database.database_config.getMetadata(group_by_field);
+      if (metadata.name == database.database_config.schema.primary_key) {
+         throw QueryParseException("Cannot group by primary key field: '" + group_by_field + "'");
+      }
+      // TODO(taepper) optimize when equal to partition_by field
+      // TODO(taepper) optimize single field groupby
+   }
+   std::vector<config::DatabaseMetadata> group_by_metadata;
+   for (const std::string& group_by_field : group_by_fields) {
+      const auto& metadata = database.database_config.getMetadata(group_by_field);
+      group_by_metadata.push_back(metadata);
+   }
+
+   std::vector<TupleAccessor> order_by_definition;
+   bool randomize_order = false;
+   for (const std::string& order_by_field : order_by_fields) {
+      if (order_by_field == "random()") {
+         randomize_order = true;
+         continue;
+      }
+      order_by_definition.push_back(parseOrderByField(order_by_field, group_by_metadata));
+   }
+
+   tbb::enumerable_thread_specific<std::unordered_map<Tuple, uint32_t>> maps;
+
+   tbb::parallel_for(
+      tbb::blocked_range<uint32_t>(0, database.partitions.size()),
+      [&](tbb::blocked_range<uint32_t> range) {
+         std::unordered_map<Tuple, uint32_t>& map = maps.local();
+         for (uint32_t partition_id = range.begin(); partition_id != range.end(); ++partition_id) {
+            const DatabasePartition& partition = database.partitions[partition_id];
+            for (const uint32_t sequence_id : *bitmap_filters[partition_id]) {
+               ++map[Tuple(sequence_id, partition.meta_store, group_by_metadata)];
+            }
+         }
+      }
+   );
+   std::unordered_map<Tuple, uint32_t> final_map;
+   for (auto& map : maps) {
+      for (auto& [key, value] : map) {
+         final_map[key] += value;
+      }
+   }
+   std::vector<Tuple> tuples;
+   for (auto& [tuple, count] : final_map) {
+      tuples.push_back(tuple);
+      tuples.back().addField(count);
+   }
+   group_by_metadata.push_back({"count", config::ValueType::NUMBER, false});
+   if (randomize_order) {
+      // obtain a time-based seed:
+      const uint32_t seed = std::chrono::system_clock::now().time_since_epoch().count();
+      std::default_random_engine rng(seed);
+      std::shuffle(tuples.begin(), tuples.end(), rng);
+   }
+   if (!order_by_definition.empty()) {
+      applyOrderBy(tuples, order_by_definition);
+   }
```

Comment:
The code read like it's possible to randomize AND order by. Is that correct?

Reply:
Yes, this comes for free and offers great functionality if the user wants to shuffle within key ranges